In [1]:
import glob
import pandas as pd
import numpy as np
from numpy import load
import torch

/media/SSD/NDT34-venv/lib/python3.8/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


# Gradients of Dataset

In [4]:
noiseGradient = torch.load('checkpoints/imdb/imdb_bert_train_noise_20%/noise20%.grad')

In [5]:
print("Tổng số phần tử trong dữ liệu:", len(noiseGradient))
print("Mỗi phần tử là list gồm {} phần tử tương ứng là w và b của last layer".format(len(noiseGradient[0])))
print("Size W: {}, Size bias: {}".format(noiseGradient[0][0].size(), noiseGradient[0][1].size()))

Tổng số phần tử trong dữ liệu: 17500
Mỗi phần tử là list gồm 2 phần tử tương ứng là w và b của last layer
Size W: torch.Size([2, 768]), Size bias: torch.Size([2])


In [6]:
classGradient = torch.load('checkpoints/imdb/imdb_bert_train_noise_20%/subclass100.grad')

In [7]:
len(classGradient)

200

# Dataset IMDB

In [2]:
data = pd.read_csv('./data/imdb/processed/noise/noise20%.csv')
data

review  label  isFlipped  \
0      i bought this movie from a market stall three ...      0          0   
1      beautifully filmed mind expanding exploration ...      1          0   
2      this movie has more goofs than any other movie...      1          1   
3      i would love to have that two hours of my life...      0          0   
4      the minute the forward started i knew we were ...      1          1   
...                                                  ...    ...        ...   
17495  dead or alive final the movie that supposedly ...      1          0   
17496  first of all before i start my review i just r...      1          0   
17497  a lot of people don't think branagh's hamlet f...      1          0   
17498  australia's first mainstream slasher film hits...      1          0   
17499  as i was reading through the comments i was su...      1          0   

       originLabel  
0                0  
1                1  
2                0  
3                0  
4                0  
...            ...  
17495            1  
17496            1  
17497            1  
17498            1  
17499            1  

[17500 rows x 4 columns]

In [3]:
subclass = pd.read_csv('./data/imdb/processed/subclass/subclass100.csv')
subclass

review  label
0    okay the story makes no sense the characters l...      0
1    t rama rao made some extremely beautiful films...      0
2    macbeth is one of the most frequently told sto...      0
3    recherche is a good word to describe this movi...      0
4    this movie is s l o w spent most of the movie ...      0
..                                                 ...    ...
195  i was ten years old when i saw subspecies i in...      1
196  although cinderella isn't the obvious choice f...      1
197  jimmy stewart and anthony mann teamed to do so...      1
198  kids of whatever age do not want to know about...      1
199  "joe" is one of those movies where although yo...      1

[200 rows x 2 columns]

# GDClass

In [11]:
from influencer.GD import GD

In [13]:
results = GD(noiseGradient, classGradient)

100%|██████████| 200/200 [01:20<00:00,  2.49it/s]


In [16]:
GDClass = pd.DataFrame(results)
GDClass

0         1          2         3          4         5         6      \
0    2.138776 -0.583733  -3.040747  2.220021  -4.369860  1.552840 -1.410145   
1    0.149659 -0.183284  -0.446987  0.233089  -0.607113  0.228392 -0.106383   
2    2.138687 -1.471097 -12.396529  3.920047  -8.466908  3.643589 -2.173310   
3    3.658803 -1.906485  -9.133389  9.769155 -11.425154  4.598601 -4.208242   
4    0.855475 -0.479705  -2.476811  1.385576  -6.171425  1.214637 -0.828342   
..        ...       ...        ...       ...        ...       ...       ...   
195 -3.237493  2.580176  10.998296 -6.255586  11.566450 -5.660321  3.999774   
196 -4.419428  3.335465  12.828603 -7.754198  14.965626 -6.295300  5.947212   
197 -1.418526  1.010637   4.944467 -2.729644   4.916633 -2.545873  2.165380   
198 -3.649356  3.569035  11.972286 -6.621225  14.783200 -7.018753  4.923038   
199 -2.603586  2.933234   7.329040 -5.056418  11.126493 -5.539310  4.168729   

        7          8         9      ...     17490     17491     17492  \
0    1.638075   2.287705 -1.415372  ...  0.284311  0.354416 -1.761626   
1    0.223117   0.300089 -0.203223  ...  0.044594  0.044585 -0.223520   
2    4.022087   5.107981 -3.577752  ...  0.639505  0.632776 -2.726520   
3    5.441940   6.764110 -4.741738  ...  0.802838  0.961955 -4.696334   
4    1.113324   1.478909 -0.889593  ...  0.210848  0.255342 -1.441293   
..        ...        ...       ...  ...       ...       ...       ...   
195 -6.922566  -8.566452  6.753926  ... -0.937020 -0.933650  4.297963   
196 -8.725953 -11.507199  6.994238  ... -0.893898 -0.973101  5.671482   
197 -3.350060  -4.257018  3.124533  ... -0.394156 -0.404637  2.382878   
198 -7.462416  -9.356909  6.521062  ... -1.166792 -1.268857  6.788964   
199 -5.464995  -7.497009  4.811496  ... -0.809739 -0.843040  4.642165   

        17493     17494      17495     17496     17497     17498      17499  
0    2.160257  0.550634  -2.053551 -1.118400 -1.127274 -0.580638  -2.337943  
1    0.231412  0.090629  -0.263697 -0.113635 -0.206221 -0.075721  -0.429619  
2    3.839478  1.166098  -5.160079 -2.434449 -3.191204 -1.061952  -5.776847  
3    6.503849  1.382574  -6.075753 -2.726885 -3.630112 -1.746696  -7.645941  
4    1.492053  0.450738  -1.507575 -0.591234 -1.168343 -0.480060  -1.849464  
..        ...       ...        ...       ...       ...       ...        ...  
195 -5.663622 -1.848246   9.170475  4.460175  4.962909  1.665207  10.422771  
196 -6.719276 -2.440602  10.359165  5.521961  5.965205  2.491141  12.267212  
197 -2.373559 -0.811574   3.824457  2.362502  2.919771  0.917836   4.453814  
198 -6.315349 -2.247442   9.150716  4.386090  7.135773  2.445084  11.318412  
199 -4.160357 -1.940008   6.809116  3.774876  5.702859  2.212487  10.604789  

[200 rows x 17500 columns]

In [18]:
top = [0.05, 0.1, 0.15, 0.2]
top = [int(len(data) * i) for i in top]
num_class = 2
step = 100

In [30]:
print("Result for GD:")
for t in top:
    first = (data.iloc[np.argsort(GDClass.sum().values)]['isFlipped'][:t] == 1).sum()
    print("Number elements: {:5} - ACC: {:.2f} %".format(t, first/t*100))

Result for GD:
Number elements:   875 - ACC: 80.80 %
Number elements:  1750 - ACC: 71.20 %
Number elements:  2625 - ACC: 58.06 %
Number elements:  3500 - ACC: 46.83 %


/tmp/ipykernel_103087/3959748290.py:3: FutureWarning: The behavior of `series[i:j]` with an integer-dtype index is deprecated. In a future version, this will be treated as *label-based* indexing, consistent with e.g. `series[i]` lookups. To retain the old behavior, use `series.iloc[i:j]`. To get the future behavior, use `series.loc[i:j]`.
  first = (data.iloc[np.argsort(GDClass.sum().values)]['isFlipped'][:t] == 1).sum()


In [19]:
df_class = pd.DataFrame()
i = 0
while i< num_class:
    class_scores = GDClass.iloc[i*step:(i+1)*step + 1].sum()
    df_class[i] = class_scores.values
    i+=1

In [20]:
df_class

0            1
0      152.789752  -271.353698
1     -102.141280   237.649323
2     -430.729385   785.023667
3      276.324152  -517.930903
4     -591.372490  1038.137325
...           ...          ...
17495 -304.021818   627.595766
17496 -147.435313   344.420956
17497 -215.329805   461.289342
17498  -86.646794   185.942815
17499 -389.578643   877.423463

[17500 rows x 2 columns]

In [21]:
scores = df_class.min(axis=1).values
scores

array([-271.35369843, -102.14127968, -430.72938529, ..., -215.3298054 ,
        -86.64679387, -389.57864317])

In [31]:
print("Result for GDClass:")
for t in top:
    first = (data.iloc[np.argsort(scores)]['isFlipped'][:t] == 1).sum()
    print("Number elements: {:5} - ACC: {:.2f} %".format(t, first/t*100))

Result for GDClass:
Number elements:   875 - ACC: 77.94 %
Number elements:  1750 - ACC: 73.83 %
Number elements:  2625 - ACC: 72.84 %
Number elements:  3500 - ACC: 69.51 %


/tmp/ipykernel_103087/2299715740.py:3: FutureWarning: The behavior of `series[i:j]` with an integer-dtype index is deprecated. In a future version, this will be treated as *label-based* indexing, consistent with e.g. `series[i]` lookups. To retain the old behavior, use `series.iloc[i:j]`. To get the future behavior, use `series.loc[i:j]`.
  first = (data.iloc[np.argsort(scores)]['isFlipped'][:t] == 1).sum()


# GDClass + Normalize

In [55]:
classGradientFlat = torch.as_tensor(np.array([torch.cat([torch.nn.functional.normalize(x.view(-1), p=2.0, dim = 0) for x in g]).cpu().numpy() for g in classGradient]))
noiseGradientFlat = torch.as_tensor(np.array([torch.cat([x.view(-1) for x in g]).cpu().numpy() for g in noiseGradient]))

In [56]:
print(classGradientFlat.shape, noiseGradientFlat.shape)

torch.Size([200, 1538]) torch.Size([17500, 1538])


In [57]:
GDClassNormalize = pd.DataFrame(torch.tensordot(classGradientFlat, noiseGradientFlat, dims = [[1], [1]]))
GDClassNormalize

0         1         2         3         4         5         6      \
0    2.210324 -0.773522 -3.564220  2.466262 -4.982144  1.831583 -1.635790   
1    1.230937 -1.408915 -3.653963  1.941172 -4.889194  1.878039 -1.036335   
2    1.240435 -0.905099 -6.377244  2.220315 -4.833453  2.062865 -1.336835   
3    1.504650 -0.883973 -3.872726  3.725303 -4.858219  1.967271 -1.769705   
4    1.305493 -0.821482 -3.782914  2.121259 -8.404491  1.883607 -1.355745   
..        ...       ...       ...       ...       ...       ...       ...   
195 -1.187428  0.967363  3.896876 -2.213748  4.273326 -2.013537  1.481561   
196 -1.231809  0.962118  3.576652 -2.130610  4.253426 -1.782241  1.641603   
197 -1.062910  0.800907  3.546779 -1.968651  3.743546 -1.833840  1.557307   
198 -1.131625  1.077430  3.608773 -2.005798  4.482430 -2.065926  1.509557   
199 -1.031009  1.090848  2.927667 -1.919279  4.242142 -2.025876  1.563631   

        7         8         9      ...     17490     17491     17492  \
0    2.101892  2.770451 -1.763497  ...  0.321614  0.395513 -2.048580   
1    2.035103  2.606526 -1.772941  ...  0.347984  0.358519 -1.875589   
2    2.424221  2.993895 -2.088498  ...  0.351718  0.359352 -1.680221   
3    2.442380  2.966903 -2.070824  ...  0.333387  0.397226 -2.028872   
4    1.958576  2.478160 -1.552957  ...  0.317255  0.381098 -2.182699   
..        ...       ...       ...  ...       ...       ...       ...   
195 -2.557845 -3.101072  2.375103  ... -0.327299 -0.336927  1.659784   
196 -2.495962 -3.178601  1.987540  ... -0.257017 -0.284600  1.675448   
197 -2.467648 -3.052920  2.209619  ... -0.282486 -0.298663  1.770903   
198 -2.358576 -2.883630  2.003726  ... -0.337163 -0.371558  2.040358   
199 -2.193767 -2.863175  1.869896  ... -0.298347 -0.318508  1.794731   

        17493     17494     17495     17496     17497     17498     17499  
0    2.438865  0.654679 -2.596169 -1.400210 -1.458078 -0.685156 -2.876220  
1    1.956542  0.732354 -2.411273 -1.109182 -1.746748 -0.641003 -3.489383  
2    2.210342  0.680167 -3.053497 -1.480858 -1.869641 -0.636877 -3.353799  
3    2.662929  0.620078 -2.774032 -1.302953 -1.654108 -0.744960 -3.321590  
4    2.274723  0.697027 -2.549275 -1.100931 -1.838369 -0.735426 -2.980457  
..        ...       ...       ...       ...       ...       ...       ...  
195 -2.073643 -0.674532  3.306499  1.637428  1.827481  0.626576  3.681334  
196 -1.932443 -0.682086  2.965854  1.576023  1.720339  0.696373  3.401772  
197 -1.795760 -0.604498  2.847222  1.693189  2.051438  0.666991  3.220669  
198 -1.959945 -0.681121  2.871045  1.410322  2.102859  0.730772  3.401982  
199 -1.685973 -0.716458  2.700034  1.476308  2.082741  0.798648  3.821258  

[200 rows x 17500 columns]

In [58]:
top = [0.05, 0.1, 0.15, 0.2]
top = [int(len(data) * i) for i in top]
num_class = 2
step = 100

df_class = pd.DataFrame()
i = 0
while i< num_class:
    class_scores = GDClassNormalize.iloc[i*step:(i+1)*step + 1].sum()
    df_class[i] = class_scores.values
    i+=1

df_class

0           1
0      127.489708 -117.707886
1      -86.562233  102.516830
2     -366.397858  338.036285
3      218.426529 -216.347336
4     -487.661835  442.976257
...           ...         ...
17495 -265.310150  278.473328
17496 -129.271454  152.651627
17497 -176.845398  193.845459
17498  -67.766861   77.192245
17499 -316.548828  362.114349

[17500 rows x 2 columns]

In [59]:
scores = df_class.min(axis=1).values
scores

array([-117.707886,  -86.56223 , -366.39786 , ..., -176.8454  ,
        -67.76686 , -316.54883 ], dtype=float32)

In [60]:
print("Result for GDClassNormalize:")
for t in top:
    first = (data.iloc[np.argsort(scores)]['isFlipped'][:t] == 1).sum()
    print("Number elements: {:5} - ACC: {:.2f} %".format(t, first/t*100))

Result for GDClassNormalize:
Number elements:   875 - ACC: 91.20 %
Number elements:  1750 - ACC: 83.60 %
Number elements:  2625 - ACC: 75.54 %
Number elements:  3500 - ACC: 67.94 %


/tmp/ipykernel_111239/1936824095.py:3: FutureWarning: The behavior of `series[i:j]` with an integer-dtype index is deprecated. In a future version, this will be treated as *label-based* indexing, consistent with e.g. `series[i]` lookups. To retain the old behavior, use `series.iloc[i:j]`. To get the future behavior, use `series.loc[i:j]`.
  first = (data.iloc[np.argsort(scores)]['isFlipped'][:t] == 1).sum()


# KNN

In [8]:
# Load features
noiseFeatures = torch.load('./checkpoints/imdb/imdb_bert_train_noise_20%/noiseFeatures.ft')
subclassFeatures = torch.load('./checkpoints/imdb/imdb_bert_train_noise_20%/subclassFeatures.ft')

In [9]:
print(len(noiseFeatures), len(subclassFeatures))

17500 200


In [10]:
from scipy.spatial import distance
sim_func_dict = {
    "dot": lambda x, y: np.dot(x, y), # càng nhỏ càng xa
    "cos": lambda x, y: distance.cosine(x, y), # càng nhỏ càng xa
    "euc": lambda x, y: -np.linalg.norm(x-y) # càng nhỏ càng xa
}


In [11]:
import tqdm
eucTable = np.zeros((len(noiseFeatures), len(noiseFeatures)))
for p, z_t in enumerate(tqdm.tqdm(noiseFeatures)):
    for q, z in enumerate(noiseFeatures):
        influence = sim_func_dict["euc"](z_t, z)
        eucTable[p][q] = influence

100%|██████████| 17500/17500 [20:01<00:00, 14.57it/s]


In [12]:
eucTable = pd.DataFrame(eucTable)
eucTable

0         1         2         3         4         5         6      \
0     -0.000000 -2.456073 -0.281153 -0.950150 -0.501907 -0.858129 -1.994658   
1     -2.456073 -0.000000 -2.631002 -1.651292 -2.795233 -1.717608 -0.584060   
2     -0.281153 -2.631002 -0.000000 -1.137756 -0.475079 -1.026376 -2.175419   
3     -0.950150 -1.651292 -1.137756 -0.000000 -1.310120 -0.675040 -1.264624   
4     -0.501907 -2.795233 -0.475079 -1.310120 -0.000000 -1.143875 -2.325175   
...         ...       ...       ...       ...       ...       ...       ...   
17495 -1.157552 -1.411315 -1.327831 -0.674091 -1.463564 -0.410073 -0.966625   
17496 -1.846581 -0.715662 -2.018637 -1.154512 -2.184849 -1.102558 -0.274928   
17497 -1.644471 -0.943037 -1.818668 -1.027578 -1.950782 -0.839222 -0.489314   
17498 -2.420506 -0.514835 -2.601488 -1.694452 -2.727012 -1.622259 -0.548594   
17499 -2.024456 -0.477786 -2.206921 -1.252361 -2.368786 -1.293676 -0.270520   

          7         8         9      ...     17490     17491     17492  \
0     -1.863564 -1.438074 -2.232023  ... -0.383818 -0.510590 -1.110047   
1     -0.635568 -1.098856 -0.362732  ... -2.634971 -2.913623 -1.389506   
2     -2.046078 -1.603456 -2.407184  ... -0.347782 -0.405858 -1.281478   
3     -1.093002 -0.684080 -1.428733  ... -1.202769 -1.374343 -0.465952   
4     -2.211165 -1.810548 -2.578036  ... -0.326308 -0.374936 -1.468536   
...         ...       ...       ...  ...       ...       ...       ...   
17495 -0.856150 -0.612635 -1.217075  ... -1.287737 -1.587622 -0.358026   
17496 -0.312269 -0.603836 -0.585554  ... -2.007344 -2.307325 -0.810962   
17497 -0.463495 -0.636829 -0.781735  ... -1.774936 -2.098871 -0.668546   
17498 -0.705163 -1.196559 -0.580711  ... -2.558887 -2.873336 -1.378279   
17499 -0.231465 -0.716311 -0.326436  ... -2.205680 -2.485666 -0.972391   

          17493     17494     17495     17496     17497     17498     17499  
0     -0.406964 -1.173543 -1.157552 -1.846581 -1.644471 -2.420506 -2.024456  
1     -2.793778 -1.463591 -1.411315 -0.715662 -0.943037 -0.514835 -0.477786  
2     -0.291315 -1.357203 -1.327831 -2.018637 -1.818668 -2.601488 -2.206921  
3     -1.248476 -0.722892 -0.674091 -1.154512 -1.027578 -1.694452 -1.252361  
4     -0.401406 -1.457487 -1.463564 -2.184849 -1.950782 -2.727012 -2.368786  
...         ...       ...       ...       ...       ...       ...       ...  
17495 -1.480512 -0.418135 -0.000000 -0.810361 -0.577064 -1.317586 -0.997345  
17496 -2.199095 -0.929000 -0.810361 -0.000000 -0.399100 -0.682662 -0.350127  
17497 -1.987354 -0.632109 -0.577064 -0.399100 -0.000000 -0.825467 -0.567651  
17498 -2.763065 -1.349199 -1.317586 -0.682662 -0.825467 -0.000000 -0.583864  
17499 -2.370003 -1.062495 -0.997345 -0.350127 -0.567651 -0.583864 -0.000000  

[17500 rows x 17500 columns]

In [25]:
eucTable[0]

0       -0.000000
1       -2.456073
2       -0.281153
3       -0.950150
4       -0.501907
           ...   
17495   -1.157552
17496   -1.846581
17497   -1.644471
17498   -2.420506
17499   -2.024456
Name: 0, Length: 17500, dtype: float64

In [70]:
numSample = len(eucTable)
k = 10
for threshold in [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]:
# threshold = 0.7
    truePositive, numEle = 0, 0
    for i in tqdm.tqdm(range(numSample)):
        # Chọn k phần tử gần nhất (trừ chính nó)
        kNearest = np.argsort(eucTable[i]).tolist()[-2:-(k+2):-1]
        # Kiểm tra sai khác label
        ilabel = int(data.loc[i, 'label'])
        iFlip = int(data.loc[i, 'isFlipped'])
        labels = data.loc[kNearest, 'label'].tolist()
        numDiff = np.sum(np.where(np.array(labels) != ilabel, 1, 0))
        # Nếu số labels nearest khác biệt lớn hơn ngưỡng
        if numDiff > threshold * k:
            numEle += 1
            if iFlip == 1:
                truePositive += 1
    print(threshold, truePositive, numEle, truePositive/numEle*100)


100%|██████████| 17500/17500 [00:31<00:00, 553.24it/s]


0.1 3403 12201 27.89115646258503


100%|██████████| 17500/17500 [00:31<00:00, 552.31it/s]


0.2 3296 8712 37.83287419651056


100%|██████████| 17500/17500 [00:31<00:00, 549.70it/s]


0.3 3140 6294 49.88878296790594


100%|██████████| 17500/17500 [00:31<00:00, 550.76it/s]


0.4 2944 4769 61.73201929125602


100%|██████████| 17500/17500 [00:31<00:00, 552.01it/s]


0.5 2659 3747 70.96343741659994


100%|██████████| 17500/17500 [00:31<00:00, 551.83it/s]


0.6 2266 2901 78.11099620820407


100%|██████████| 17500/17500 [00:31<00:00, 551.46it/s]


0.7 1676 2002 83.71628371628371


100%|██████████| 17500/17500 [00:31<00:00, 552.57it/s]


0.8 942 1081 87.14153561517114


100%|██████████| 17500/17500 [00:31<00:00, 551.48it/s]

0.9 307 345 88.98550724637681


k = 10

threashold = 0.1            Acc = 27.89

threashold = 0.2            Acc = 37.83

threashold = 0.3            Acc = 49.89

threashold = 0.4            Acc = 61.73

threashold = 0.5            Acc = 70.96

threashold = 0.6            Acc = 78.11

threashold = 0.7            Acc = 83.72

threashold = 0.8            Acc = 87.14

threashold = 0.9           Acc = 88.98

# KNN + Normalize Gradient

Với mỗi sample $z$, Chọn ra $k$ điểm gần nhất bằng KNN dựa trên features (Các điểm này gọi là $z'$), sau đó tính gradient của $k$ điểm $z'$ và normalize chúng, tính GD của $k$ điểm $z'$ này với điểm $z$. Score cuối của $z$ chính là tổng các GD của $z$ với $z'$. 

In [43]:
eucTable

0         1         2         3         4         5         6      \
0     -0.000000 -2.456073 -0.281153 -0.950150 -0.501907 -0.858129 -1.994658   
1     -2.456073 -0.000000 -2.631002 -1.651292 -2.795233 -1.717608 -0.584060   
2     -0.281153 -2.631002 -0.000000 -1.137756 -0.475079 -1.026376 -2.175419   
3     -0.950150 -1.651292 -1.137756 -0.000000 -1.310120 -0.675040 -1.264624   
4     -0.501907 -2.795233 -0.475079 -1.310120 -0.000000 -1.143875 -2.325175   
...         ...       ...       ...       ...       ...       ...       ...   
17495 -1.157552 -1.411315 -1.327831 -0.674091 -1.463564 -0.410073 -0.966625   
17496 -1.846581 -0.715662 -2.018637 -1.154512 -2.184849 -1.102558 -0.274928   
17497 -1.644471 -0.943037 -1.818668 -1.027578 -1.950782 -0.839222 -0.489314   
17498 -2.420506 -0.514835 -2.601488 -1.694452 -2.727012 -1.622259 -0.548594   
17499 -2.024456 -0.477786 -2.206921 -1.252361 -2.368786 -1.293676 -0.270520   

          7         8         9      ...     17490     17491     17492  \
0     -1.863564 -1.438074 -2.232023  ... -0.383818 -0.510590 -1.110047   
1     -0.635568 -1.098856 -0.362732  ... -2.634971 -2.913623 -1.389506   
2     -2.046078 -1.603456 -2.407184  ... -0.347782 -0.405858 -1.281478   
3     -1.093002 -0.684080 -1.428733  ... -1.202769 -1.374343 -0.465952   
4     -2.211165 -1.810548 -2.578036  ... -0.326308 -0.374936 -1.468536   
...         ...       ...       ...  ...       ...       ...       ...   
17495 -0.856150 -0.612635 -1.217075  ... -1.287737 -1.587622 -0.358026   
17496 -0.312269 -0.603836 -0.585554  ... -2.007344 -2.307325 -0.810962   
17497 -0.463495 -0.636829 -0.781735  ... -1.774936 -2.098871 -0.668546   
17498 -0.705163 -1.196559 -0.580711  ... -2.558887 -2.873336 -1.378279   
17499 -0.231465 -0.716311 -0.326436  ... -2.205680 -2.485666 -0.972391   

          17493     17494     17495     17496     17497     17498     17499  
0     -0.406964 -1.173543 -1.157552 -1.846581 -1.644471 -2.420506 -2.024456  
1     -2.793778 -1.463591 -1.411315 -0.715662 -0.943037 -0.514835 -0.477786  
2     -0.291315 -1.357203 -1.327831 -2.018637 -1.818668 -2.601488 -2.206921  
3     -1.248476 -0.722892 -0.674091 -1.154512 -1.027578 -1.694452 -1.252361  
4     -0.401406 -1.457487 -1.463564 -2.184849 -1.950782 -2.727012 -2.368786  
...         ...       ...       ...       ...       ...       ...       ...  
17495 -1.480512 -0.418135 -0.000000 -0.810361 -0.577064 -1.317586 -0.997345  
17496 -2.199095 -0.929000 -0.810361 -0.000000 -0.399100 -0.682662 -0.350127  
17497 -1.987354 -0.632109 -0.577064 -0.399100 -0.000000 -0.825467 -0.567651  
17498 -2.763065 -1.349199 -1.317586 -0.682662 -0.825467 -0.000000 -0.583864  
17499 -2.370003 -1.062495 -0.997345 -0.350127 -0.567651 -0.583864 -0.000000  

[17500 rows x 17500 columns]

In [44]:
noiseGradientFlat = torch.as_tensor(np.array([torch.cat([x.view(-1) for x in g]).cpu().numpy() for g in noiseGradient]))

In [45]:
numSample = len(eucTable)
k = 100
scores = []
for i in tqdm.tqdm(range(numSample)):
    # Chọn k phần tử gần nhất (Trừ chính nó)
    kNearest = np.argsort(eucTable[i]).tolist()[-2:-(k+2):-1]
    # Chọn k graident và normalize k gradient này
    kGradient = noiseGradientFlat[kNearest]
    kGradientNormalize = torch.as_tensor(np.array([torch.nn.functional.normalize(x, p=2.0, dim=0).numpy() for x in kGradient]))
    # Tính Score cho điểm đó
    GDKnnNormalize = torch.tensordot(kGradientNormalize, noiseGradientFlat[i], dims = [[1], [0]]).sum().item()
    # Lưu score lại
    scores.append(GDKnnNormalize)
    

100%|██████████| 17500/17500 [00:55<00:00, 316.54it/s]


In [46]:
top = [0.05, 0.1, 0.15, 0.2]
top = [int(len(data) * i) for i in top]
print("Result for GD:")
for t in top:
    first = (data.iloc[np.argsort(scores)]['isFlipped'][:t] == 1).sum()
    print("Number elements: {:5} - ACC: {:.2f} %".format(t, first/t*100))

Result for GD:
Number elements:   875 - ACC: 93.83 %
Number elements:  1750 - ACC: 89.66 %
Number elements:  2625 - ACC: 85.64 %
Number elements:  3500 - ACC: 78.69 %


/tmp/ipykernel_2925820/3513638168.py:5: FutureWarning: The behavior of `series[i:j]` with an integer-dtype index is deprecated. In a future version, this will be treated as *label-based* indexing, consistent with e.g. `series[i]` lookups. To retain the old behavior, use `series.iloc[i:j]`. To get the future behavior, use `series.loc[i:j]`.
  first = (data.iloc[np.argsort(scores)]['isFlipped'][:t] == 1).sum()


Out ra cả số phần tử

# KNN + Class + Normalize

Với mỗi sample $z$, Chọn ra $k$ điểm gần nhất trên mỗi class bằng KNN dựa trên features (Các điểm này gọi là $z'$ => Có class x $k$ phần tử được chọn), sau đó tính gradient của $k$ điểm $z'$ mỗi class và normalize chúng, tính GD của $k$ điểm $z'$ mỗi class với điểm $z$, thu được $class$ scores. Score cuối của $z$ chính là min các score trên class.

In [47]:
eucTable

0         1         2         3         4         5         6      \
0     -0.000000 -2.456073 -0.281153 -0.950150 -0.501907 -0.858129 -1.994658   
1     -2.456073 -0.000000 -2.631002 -1.651292 -2.795233 -1.717608 -0.584060   
2     -0.281153 -2.631002 -0.000000 -1.137756 -0.475079 -1.026376 -2.175419   
3     -0.950150 -1.651292 -1.137756 -0.000000 -1.310120 -0.675040 -1.264624   
4     -0.501907 -2.795233 -0.475079 -1.310120 -0.000000 -1.143875 -2.325175   
...         ...       ...       ...       ...       ...       ...       ...   
17495 -1.157552 -1.411315 -1.327831 -0.674091 -1.463564 -0.410073 -0.966625   
17496 -1.846581 -0.715662 -2.018637 -1.154512 -2.184849 -1.102558 -0.274928   
17497 -1.644471 -0.943037 -1.818668 -1.027578 -1.950782 -0.839222 -0.489314   
17498 -2.420506 -0.514835 -2.601488 -1.694452 -2.727012 -1.622259 -0.548594   
17499 -2.024456 -0.477786 -2.206921 -1.252361 -2.368786 -1.293676 -0.270520   

          7         8         9      ...     17490     17491     17492  \
0     -1.863564 -1.438074 -2.232023  ... -0.383818 -0.510590 -1.110047   
1     -0.635568 -1.098856 -0.362732  ... -2.634971 -2.913623 -1.389506   
2     -2.046078 -1.603456 -2.407184  ... -0.347782 -0.405858 -1.281478   
3     -1.093002 -0.684080 -1.428733  ... -1.202769 -1.374343 -0.465952   
4     -2.211165 -1.810548 -2.578036  ... -0.326308 -0.374936 -1.468536   
...         ...       ...       ...  ...       ...       ...       ...   
17495 -0.856150 -0.612635 -1.217075  ... -1.287737 -1.587622 -0.358026   
17496 -0.312269 -0.603836 -0.585554  ... -2.007344 -2.307325 -0.810962   
17497 -0.463495 -0.636829 -0.781735  ... -1.774936 -2.098871 -0.668546   
17498 -0.705163 -1.196559 -0.580711  ... -2.558887 -2.873336 -1.378279   
17499 -0.231465 -0.716311 -0.326436  ... -2.205680 -2.485666 -0.972391   

          17493     17494     17495     17496     17497     17498     17499  
0     -0.406964 -1.173543 -1.157552 -1.846581 -1.644471 -2.420506 -2.024456  
1     -2.793778 -1.463591 -1.411315 -0.715662 -0.943037 -0.514835 -0.477786  
2     -0.291315 -1.357203 -1.327831 -2.018637 -1.818668 -2.601488 -2.206921  
3     -1.248476 -0.722892 -0.674091 -1.154512 -1.027578 -1.694452 -1.252361  
4     -0.401406 -1.457487 -1.463564 -2.184849 -1.950782 -2.727012 -2.368786  
...         ...       ...       ...       ...       ...       ...       ...  
17495 -1.480512 -0.418135 -0.000000 -0.810361 -0.577064 -1.317586 -0.997345  
17496 -2.199095 -0.929000 -0.810361 -0.000000 -0.399100 -0.682662 -0.350127  
17497 -1.987354 -0.632109 -0.577064 -0.399100 -0.000000 -0.825467 -0.567651  
17498 -2.763065 -1.349199 -1.317586 -0.682662 -0.825467 -0.000000 -0.583864  
17499 -2.370003 -1.062495 -0.997345 -0.350127 -0.567651 -0.583864 -0.000000  

[17500 rows x 17500 columns]

In [48]:
noiseGradientFlat = torch.as_tensor(np.array([torch.cat([x.view(-1) for x in g]).cpu().numpy() for g in noiseGradient]))

In [65]:
numSample = len(eucTable)
numClass = 2
k = 100
scores = []
Nearest = np.argsort(eucTable[i]).tolist()[-2::-1] 
for i in tqdm.tqdm(range(numSample)):
    iscore = np.inf
    for c in range(numClass):
        # Chọn k điểm ở class c gần với điểm z nhất
        indexByClass = []
        for ind in Nearest:
            if data.loc[ind, 'label'] == c:
                indexByClass.append(ind)
            if len(indexByClass) == 100:
                break
        # Lấy gradient và normalize
        kGradient = noiseGradientFlat[indexByClass]
        kGradientNormalize = torch.as_tensor(np.array([torch.nn.functional.normalize(x, p=2.0, dim=0).numpy() for x in kGradient]))
        # Tính score trên class này
        GDKnnNormalize = torch.tensordot(kGradientNormalize, noiseGradientFlat[i], dims = [[1], [0]]).sum().item()
        if GDKnnNormalize < iscore:
            iscore = GDKnnNormalize
    scores.append(iscore)
    

100%|██████████| 17500/17500 [02:21<00:00, 123.59it/s]


In [66]:
top = [0.05, 0.1, 0.15, 0.2]
top = [int(len(data) * i) for i in top]
print("Result for GD:")
for t in top:
    first = (data.iloc[np.argsort(scores)]['isFlipped'][:t] == 1).sum()
    print("Number elements: {:5} - ACC: {:.2f} %".format(t, first/t*100))

Result for GD:
Number elements:   875 - ACC: 80.57 %
Number elements:  1750 - ACC: 75.94 %
Number elements:  2625 - ACC: 70.06 %
Number elements:  3500 - ACC: 64.20 %


/tmp/ipykernel_2925820/3513638168.py:5: FutureWarning: The behavior of `series[i:j]` with an integer-dtype index is deprecated. In a future version, this will be treated as *label-based* indexing, consistent with e.g. `series[i]` lookups. To retain the old behavior, use `series.iloc[i:j]`. To get the future behavior, use `series.loc[i:j]`.
  first = (data.iloc[np.argsort(scores)]['isFlipped'][:t] == 1).sum()
